### DISCUSSION TITLE
What types of metadata can I add to a Dagster asset?

### DISCUSSION THREAD

**Discussion by slopp** [dagster-io/dagster#15409] [Q&A] [answered]
Many dagster examples show asset metadata like the asset owner, the number of rows or execution time plotted, or the table schema. How can I add these values for my own asset?

**Comment by slopp** [comment, marked as answer]
Dagster allows you to attach rich metadata in a variety of places, see this [discussion](https://github.com/dagster-io/dagster/discussions/12350).

A common approach is to add metadata to the asset materialization. Here is an example:

```python
from dagster import asset, MetadataValue, AssetExecutionContext
import random 
import pandas as pd
from dagster_pandas.data_frame import create_table_schema_metadata_from_dataframe


@asset
def my_asset(context: AssetExecutionContext):
    """ the asset description """

    data = pd.DataFrame({
        "orders": [1,2,3,4],
        "users": ["pam", "jim", "michael", "dwight"]
    })

    context.add_output_metadata({
        "numeric_value": random.choice(range(1,100)),
        "key": "value",
        "tabular_data": MetadataValue.md(data.head().to_markdown()),
        "table_schema": create_table_schema_metadata_from_dataframe(data)
    })

    return data
```

Here is the result:

![Screen Shot 2023-07-20 at 8 28 06 AM](https://github.com/dagster-io/dagster/assets/15901072/c08aed94-f62b-4964-a22a-0515e1d35aa0)

Note that numeric metadata will automatically be plotted by materialization.


Another example from our quickstart project includes adding an image: https://github.com/dagster-io/dagster/blob/master/examples/quickstart_etl/quickstart_etl/assets/hackernews.py#L67-L81


Other examples and guides from the docs:
- https://docs.dagster.io/concepts/assets/software-defined-assets#recording-materialization-metadata
- https://docs.dagster.io/guides/dagster/managing-ml#enhancing-the-dagster-ui-with-metadata
- https://docs.dagster.io/tutorial/building-an-asset-graph#step-3-educating-users-with-metadata

**Reply by ilyanekhay-uta** [reply]
The above example seems to not be working anymore; I'm getting this error:

> dagster._core.errors.DagsterInvalidDefinitionError: "context" is not a valid name in Dagster. It conflicts with a Dagster or python reserved keyword.

Would a better approach now be having the asset function return an instance of `dagster.Output`?

**Comment by Swejal08** [comment]
`  context.add_output_metadata({
        "numeric_value": random.choice(range(1,100)),
        "key": "value",
        "tabular_data": MetadataValue.md(data.head().to_markdown()),
        "table_schema": create_table_schema_metadata_from_dataframe(data)
    })`
    
   After adding this to my asset, can i access these metadatas in other assets?
